pkg/operator/controllers/azurensg/subnetnsg.go

```diff
@@ -0,0 +1,120 @@
+package azurensg
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-07-01/network"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	azureproviderv1beta1 "sigs.k8s.io/cluster-api-provider-azure/pkg/apis/azureprovider/v1beta1"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/network"
+	"github.com/Azure/ARO-RP/pkg/util/subnet"
+)
+
+type subnetDescriptor struct {
+	resourceGroup string
+	vnetName      string
+	subnetName    string
+}
+
+func (r *AzureNSGReconciler) reconcileSubnetNSG(ctx context.Context, instance *arov1alpha1.Cluster, subscriptionID string, subnetsClient network.SubnetsClient) error {
+	// the main logic starts here
+	subnets, masterResourceGroup, err := r.getSubnets(ctx)
+	if err != nil {
+		return err
+	}
+	for ws := range subnets {
+		err = r.ensureSubnetNSG(ctx, subnetsClient, subscriptionID, masterResourceGroup, instance.Spec.InfraID, api.ArchitectureVersion(instance.Spec.ArchitectureVersion), ws.resourceGroup, ws.vnetName, ws.subnetName, subnets[ws])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *AzureNSGReconciler) ensureSubnetNSG(ctx context.Context, subnetsClient network.SubnetsClient, subscriptionID, resourcesResourceGroup, infraID string, architectureVersion api.ArchitectureVersion, vnetResourceGroup, vnetName, subnetName string, isWorkerSubnet bool) error {
+	subnetObject, err := subnetsClient.Get(ctx, vnetResourceGroup, vnetName, subnetName, "")
+	if err != nil {
+		return err
+	}
+	if subnetObject.SubnetPropertiesFormat == nil {
+		return fmt.Errorf("received nil SubnetProperties when trying to Get subnet %s/%s in resource group %s", vnetName, subnetName, vnetResourceGroup)
+	}
+	if subnetObject.SubnetPropertiesFormat.NetworkSecurityGroup == nil {
+		return fmt.Errorf("received nil NSG when trying to Get subnet %s/%s in resource group %s", vnetName, subnetName, vnetResourceGroup)
+	}
```

Comment:
Is it important for us to distinguish between these two errors? Or can we combine: `if subnetObject.SubnetPropertiesFormat == nil || subnetObject.SubnetPropertiesFormat.NetworkSecurityGroup == nil`?

Also, I think we should test this case. Currently both returns are not covered.

Reply:
I merged the if.

Reply:
Added a test case.